Remove MemPoolContext

ngimel opened the pull request:
Removes MemPoolContext from custom user mempools. The ground truth for which pool should be used is in graph_pools active pool, and MemPoolContext just introduced an opportunity for the pool pointed to by MemPoolContext and active pool in graph_pools to go out of sync (see all the asserts in the code to make sure that happens, and yet it still could happen in a multithread scenario, see my recent PRs (#153990).




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

ngimel commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -1276,17 +1283,20 @@ class DeviceCachingAllocator {
           // Free all non-split cached blocks and retry alloc.
           || (C10_LIKELY(captures_underway.empty()) &&
-              release_cached_blocks(context) &&
+              release_cached_blocks(context, {0, 0}) &&
```
this is a different behavior introduced by this PR - in about to oom situation, after allocation failed, we release all that we can to release, not just the blocks in the current pool, to increase chances that retrial succeeds.

albanD commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -721,8 +721,25 @@ CapturedTraceback* getFromContext(
+    Py_ssize_t size = PyTuple_Size(arg);
+    TORCH_CHECK(size == 2, "mempool_id must be a tuple of 2 integers");
+    
+    PyObject* id1 = PyTuple_GetItem(arg, 0);
+    PyObject* id2 = PyTuple_GetItem(arg, 1);
```
You need to wrap these into THPObjectPtr otherwise you'll leak them :)

albanD commented on torch/csrc/cuda/CUDAPluggableAllocator.h:
```diff
@@ -114,7 +114,7 @@ struct TORCH_CUDA_CPP_API CUDAPluggableAllocator
   double getMemoryFraction(c10::DeviceIndex device) override;
   void setMemoryFraction(double fraction, c10::DeviceIndex device) override;
-  void emptyCache() override;
+  void emptyCache(c10::cuda::MempoolId_t) override;
```
Are these public APIs? Should we add default value to not BC-break?

albanD commented on torch/cuda/memory.py:
```diff
@@ -1177,11 +1155,8 @@ def snapshot(self):
-            snapshot = torch.cuda.memory_snapshot()
-        finally:
-            del ctx
+        # TODO fix snapshot to return this pool only
```
Isn't this done? You do thread the id all the way down to filter things?

ngimel replied:
Done indeed, will remove TODO

albanD reviewed (approved):
SGTM !

syed-ahmed reviewed (approved):
Thank you so much! LGTM!

ngimel committed:
Remove MemPoolContext (#154042)

Summary:
Removes MemPoolContext from custom user mempools. The ground truth for which pool should be used is in graph_pools active pool, and MemPoolContext just introduced an opportunity for the pool pointed to by MemPoolContext and active pool in graph_pools to go out of sync (see all the asserts in the code to make sure that happens, and yet it still could happen in a multithread scenario, see my recent PRs (https://github.com/pytorch/pytorch/issues/153990).




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k


Differential Revision: D75184810

Pulled By: ngimel

ngimel commented:
@pytorchbot merge -f "changes requested from bot"

ngimel commented:
Closing in favor of #154746
